src/Grpc.StatusProto/README.md

```diff
@@ -0,0 +1,303 @@
+# gRPC C# API for error handling with status.proto
+
+This is a protoype NuGet package providing C# and .NET client and server side support for the
+[gRPC richer error model](https://grpc.io/docs/guides/error/#richer-error-model).
+
+This feature is already available in many other implementations including C++,
+Go, Java and Python.
+
+This package has dependencies on these NuGet packages:
+* `Google.Api.CommonProtos` - to provide the proto implementations used by the richer error model
+* `Grpc.Core.Api` - for API classes such as `RpcException`
+
+## Error handling in gRPC
+
+The standard way for gRPC to report the success or failure of a gRPC call is for a
+status code to be returned. If a call completes successfully the server returns an `OK`
+status to the client, otherwise an error status code is returned with an optional string
+error message that provides further details about what happened. This is known as the
+_standard error model_ and is the official gRPC error model supported by all gRPC
+implementations.
+
+There is another error model known as the _richer error model_ that allows additional
+error details to be included by the server. These are expressed in protocol buffers
+messages, and a
+[set of standard error message types](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto)
+is defined to cover most needs. The protobuf binary encoding of this extra error
+information is provided as trailing metadata in the response.
+
+For more information on the richer error model see the
+[gRPC documentation on error handling](https://grpc.io/docs/guides/error/),
+and the [Google APIs overview of the error model](https://cloud.google.com/apis/design/errors#error_model).
+
+## .NET implementation of the richer error model
+
+The error model is define by the protocol buffers files [status.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto)
+and [error_details.proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto),
+and the `Google.Api.CommonProtos` NuGet package that provides the generated .NET classes
+from these proto files.
+
+The error is encapsulated by an instance of `Google.Rpc.Status` and
+returned in the trailing response metadata with well-known key `grpc-status-details-bin`.
+Setting and reading this metadata is handled
+for you when using the methods provided in this package.
+
+## Server Side
+
+The server side uses C#'s Object and Collection initializer syntax.
+
+The server returns the additional error information by throwing an `RpcException` that is
+created from a `Google.Rpc.Status` which contains the details of the error.
+
+To add messages to the `Details` repeated field in `Google.Rpc.Status`, wrap each one in `Any.Pack()` - see example below.
+
+The `Google.Rpc.Status` extension method `ToRpcException` creates the appropriate `RpcException` from the status.
+
+__Example__ - creating and throwing a `RpcException`:
```

Comment:
Will there be a way to use this without requiring the implementation to throw an exception?

Reply:
You can instead explicitly set the `Status` and add to the `ResponseTrailers` in the `ServerCallContext`
before returning from the call instead (this was documented in previous commit but was removed to keep it simple).  However the client will still receive an Exception since any non-OK response status generates an exception on the client.